subsys/llext/llext.c

```diff
@@ -239,33 +241,38 @@ static int llext_map_sections(struct llext_loader *ldr, struct llext *ext)
 			return ret;
 		}
 
-		/* Identify the lowest and the highest data sections */
-		if (!(shdr.sh_flags & SHF_EXECINSTR) &&
-		    shdr.sh_type == SHT_PROGBITS) {
-			if (shdr.sh_offset > high_shdr.sh_offset) {
-				high_shdr = shdr;
-			}
-			if (shdr.sh_offset < low_shdr.sh_offset) {
-				low_shdr = shdr;
-			}
+		if ((shdr.sh_type != SHT_PROGBITS && shdr.sh_type != SHT_NOBITS) ||
+		    !(shdr.sh_flags & SHF_ALLOC)) {
+			continue;
 		}
 
 		name = llext_string(ldr, ext, LLEXT_MEM_SHSTRTAB, shdr.sh_name);
 
 		LOG_DBG("section %d name %s", i, name);
 
-		enum llext_mem mem_idx;
-
-		/*
-		 * .rodata section is optional. If there isn't one, use the
-		 * first read-only data section
-		 */
-		if (shdr.sh_addr && !(shdr.sh_flags & (SHF_WRITE | SHF_EXECINSTR)) &&
-		    shdr.sh_addr < rodata.sh_addr) {
-			rodata = shdr;
-			LOG_DBG("rodata: select %#zx name %s", (size_t)shdr.sh_addr, name);
+		/* Identify the lowest and the highest data sections */
+		if (!(shdr.sh_flags & SHF_EXECINSTR)) {
+			if (shdr.sh_flags & SHF_WRITE) {
+				/* .data and other writable sections */
+				if (shdr.sh_offset > data_shdr_high.sh_offset) {
+					data_shdr_high = shdr;
+				}
+				if (shdr.sh_offset < data_shdr_low.sh_offset) {
+					data_shdr_low = shdr;
+				}
+			} else {
+				/* .rodata and others */
```

Comment:
Are we guaranteed to have these sections in a contiguous set of section headers? I'm suspicious of this being true but as always I'm not claiming expertise in ELF semantics here so I could very well be wrong.

This seems to expect that to be true though no?

Reply:
@teburd that is a correct observation. And I don't know whether that's how objects are **always** linked but so far that's what I've seen. And in fact SOF does that explicitly. We first build a generic cmake target like an executable or a shared library to compile and link all source files, but then we check what sections the resulting binary contains and re-link the individual object files while **explicitly** specifying starting address for all the sections and at that point we take care to group them accordingly: executable, writable and read-only.
That of course is just how SOF works. We could propose to use this method for building all LLEXT binaries but I'm not sure whether that would be a good option. But at least this approach isn't breaking the present situation: currently we only handle `.text`, `.data`, `.rodata`, `.bss` and `.exported_sym`. I'm not sure about the latter, we can check respective twister logs, but twister tests didn't fail.
So, I think we need to improve LLEXT's handling of "non-standard" sections and doing it in a fully generic way would be a reasonable chunk of work. This commit is a rather simplistic attempt to fix that, but I agree that "synthesising" section headers might not be very elegant. As an alternative I probably could propose 4 additional fields for `struct llext_loader`: `data_rw_start`, `data_rw_end`, `data_ro_start`, and `data_ro_end`. Would that be preferred?

Reply:
It does make sense, and I suppose would closely relate to how elf segments are done in shared and executable objects presumably. I think to accept this we need to guarantee the sections of like flags are contiguous in the elf when being mapped to sensibly do this, otherwise it’s an unwritten rule that would inadvertently fail someone somewhere.